storage/factory/common.go

```diff
@@ -117,3 +134,77 @@ func getLastEpochFromDirNames(epochDirs []string) (uint32, error) {
 
 	return epochsInDirName[0], nil
 }
+
+func getLastEpochFromStorage(
+	config config.Config,
+	marshalizer marshal.Marshalizer,
+	parentDir string,
+	defaultEpochString string,
+	defaultShardString string,
+) (uint32, error) {
+	persisterFactory := NewPersisterFactory(config.BootstrapStorage.DB)
+	pathWithoutShard := parentDir + string(os.PathSeparator) + defaultEpochString + "_0"
+	shardIdStr, err := getShardFromDirectory(pathWithoutShard, defaultShardString)
+	if err != nil {
+		return 0, err
+	}
+
+	persisterPath := pathWithoutShard + string(os.PathSeparator) + defaultShardString
+	persisterPath += shardIdStr + string(os.PathSeparator) + config.BootstrapStorage.DB.FilePath
+
+	persister, err := persisterFactory.Create(persisterPath)
+	if err != nil {
+		return 0, err
+	}
+
+	defer func() {
+		errClose := persister.Close()
+		log.LogIfError(errClose)
+	}()
+
+	cacher, err := lrucache.NewCache(10)
+	if err != nil {
+		return 0, nil
+	}
+
+	storer, err := storageUnit.NewStorageUnit(cacher, persister)
+	if err != nil {
+		return 0, err
+	}
+
+	bootStorer, err := bootstrapStorage.NewBootstrapStorer(marshalizer, storer)
+	if err != nil {
+		return 0, err
+	}
+
+	highestRound := bootStorer.GetHighestRound()
+	bootstrapData, err := bootStorer.Get(highestRound)
+	if err != nil {
+		return 0, err
+	}
+
+	log.Debug("found epoch in BootstrapData", "epoch", bootstrapData.LastHeader.Epoch)
+	return bootstrapData.LastHeader.Epoch, nil
+}
+
+func getShardFromDirectory(path string, defaultShardString string) (string, error) {
```

Comment:
what happens when the node has disabled pruning, but is a validator and gets shuffled between shards/
Would it create for each shard a new directory?
Then after a number of epochs he will have a directory for each shard. 
This implementation doesn't seem to take this into account, and will return the shardId for the first path which has a chance of only 1/(nbShards+meta) of being correct.

I think you need to search in all shard dirs for the highest round and then take the epoch from the db with highest round.

Reply:
done. Now the function searches inside all persister and returns the highest value. + the last round